File: FourthChallenge/FourthC.py
import re

def decode(message):
    regex = r'\(([^()]+)\)'
    match = re.search(regex, message)

    if not match:
        return message

    reversed_content = match.group(1)[::-1]
    replaced = re.sub(regex, reversed_content, message, 1)
    result = decode(replaced)
    return result
  
  #Explanation of Python Code:
#The Python re module is used for regular expressions.
#re.search() is used to find the first occurrence of the regex pattern in the message.
#re.sub() is utilized to replace the matched pattern with the reversed content.
#The function decode() recursively calls itself to continue decoding the message until no more 
#parentheses patterns are found.
#The logic and regex pattern remain consistent with the JavaScript code, 
# allowing this Python code to decode messages in a similar manner by reversing content enclosed within parentheses.

File: FourthChallenge/fourthC.js
function decode(message) {
    const regex = /\(([^()]+)\)/
    const match = message.match(regex)
  
    if (!match) return message 
  
    const reversed = match[1].split('').reverse().join('')
    const replaced = message.replace(regex, reversed)
    const result = decode(replaced) 
    return result 
  }

  const a = decode('hola (odnum)')
console.log(a) // hola mundo

const b = decode('(olleh) (dlrow)!')
console.log(b) // hello world!

const c = decode('sa(u(cla)atn)s')
console.log(c) // santaclaus

// Step by step:
// 1. Reverse the nested -> sa(ualcatn)s
// 2. Reverse the remaining one -> santaclaus

/**Regular Expression (Regex):

const regex = /\(([^()]+)\)/;
This regular expression searches for a pattern enclosed in parentheses. Specifically, it captures any sequence of characters that are not parentheses.
Matching:

const match = message.match(regex);
It tries to find the first match of the regex pattern in the message string.
Decoding:

if (!match) return message;
If no match is found, it means there are no parentheses left in the message, so it returns the original message.
Reverse and Replace:

const reversed = match[1].split('').reverse().join('');
It takes the content inside the parentheses from the matched pattern, splits it into individual characters, reverses the order of characters, and joins them back into a string.
const replaced = message.replace(regex, reversed);
It replaces the content inside the parentheses with the reversed content in the message string.
Recursion:

const result = decode(replaced);
It recursively calls the decode function with the updated replaced message.
The function continues this process until there are no more parentheses to decode, and it returns the final decoded message. */
